### PR TITLE
GitHub HTTP API client: support token pool, prepare for comma-separated tokens in GITHUB_API_TOKEN

### PR DESCRIPTION
**Pull request by jgehrcke**
This is for addressing #917.

### CODE REVIEW

File: conbench/entities/commit.py [hunk 1]
```diff
@@ -532,6 +534,62 @@ def __init__(self):
...
+            return
+
+        if len(tokens_to_use) == 1:
+            self._current_auth_token = tokens_to_use[0]
+            log.info("configured one GitHub HTTP API auth token")
+            return
+
+        self._token_pool = itertools.cycle(tokens_to_use)
+        self._current_auth_token = next(self._token_pool)
+        log.info("current auth token starts with: %s", self._current_auth_token[:3])
```
Comment (jgehrcke) [inline conbench/entities/commit.py]: note: use _rotate_auth_token() here instead

File: conbench/entities/commit.py [hunk 2]
```diff
@@ -820,6 +877,10 @@ def _get_response_retrycore(self, url) -> Optional[dict]:
             )
 
             if reqquota == 0:
+                if self._rotate_auth_token():
+                    log.info("rotated auth token, retry request")
+                    return None
```
Comment (jgehrcke) [inline conbench/entities/commit.py]: Note to self: need to catch case where a full rotation was performed

File: conbench/entities/commit.py [hunk 3]
```diff
@@ -532,6 +534,62 @@ def __init__(self):
...
+        self._current_auth_token = None
+        if data is None:
+            log.info("GITHUB_API_TOKEN env not set")
+            return
+
+        log.info("GITHUB_API_TOKEN env was set, length of data: %s", len(data))
+
+        token_candidates = [data]
+        if "," in data:
+            token_candidates = data.split(",")
```
Comment (austin3dickey) [inline conbench/entities/commit.py]: nit: `split()` will always return a list even if there's no comma!
```suggestion
        token_candidates = data.split(",")
```

File: conbench/entities/commit.py [hunk 4]
```diff
@@ -532,6 +534,62 @@ def __init__(self):
...
+        """
+        Return True if token was rotated, False otherwise.
+        """
+        if self._token_pool is None:
+            # convention: pool size greater than 1, rotation takes effect.
+            log.debug("tried rotating auth token, but no pool is configured")
+            return False
+
+        self._current_auth_token = next(self._token_pool)
+        log.info("current auth token starts with: %s", self._current_auth_token[:3])
```
Comment (austin3dickey) [inline conbench/entities/commit.py]: nit: I think personal access tokens always start with `ghp_`. Maybe log the last three chars instead?

File: conbench/entities/commit.py (hunk 1 above)
Comment (jgehrcke) [reply conbench/entities/commit.py]: done

File: conbench/entities/commit.py (hunk 2 above)
Comment (jgehrcke) [reply conbench/entities/commit.py]: done

File: conbench/entities/commit.py (hunk 4 above)
Comment (jgehrcke) [reply conbench/entities/commit.py]: Yes! and there also is the prefix `github_pat_` for the new fine-grained tokens. I opted for still showing prefix, just a tiny bit longer (with length depending on token type).

**Comment by jgehrcke** [comment]
Log lines from CI to show parts of this in action:
```
[conbench.entities.commit] INFO: configured GitHub HTTP API authentication token pool: ghp_BV..., ghp_e1..., ghp_i7...
[conbench.entities.commit] INFO: current auth token has length 40 and starts with: ghp_BV
```

**Comment by jgehrcke** [comment]
Token rotation wasn't hit yet in CI (according to logs), but it will probably be hit soon when I land this now.

(more review feedback welcome, and I want to really provoke the rotation code path now to see if this still contains a bug -- I even didn't hit this yet locally, misusing CI for that!).

**Comment by jgehrcke** [comment]
In first build on `main` after merging this, still no rotation performed. Will report back when I see this.

File: migrations/versions/d3515ecea53d_backfill_all_commits.py [hunk 5]
```diff
@@ -43,8 +42,11 @@ def upgrade():
             continue
 
         name = repository_to_name(repo)
-        default_branch = github.get_default_branch(name)
-        all_commits = github.get_commits_to_branch(
+
+        # This deliberately (mis)uses the per-process singleton GitHub HTTP API
+        # client object from commit.py
```
Comment (austin3dickey) [inline migrations/versions/d3515ecea53d_backfill_all_commits.py:47]: Yeah, I shouldn't have done this, but it seemed much easier than rolling my own code to do these things at the time

Comment (jgehrcke) [reply migrations/versions/d3515ecea53d_backfill_all_commits.py:47]: That's completely fine! :)

**Pull request merged**